stan/math/rev/functor/reduce_sum.hpp

```diff
@@ -0,0 +1,523 @@
+                                       Pargs&&... args) {
+      return accumulate_adjoints(dest, std::forward<Pargs>(args)...);
+    }
+
+    inline double* accumulate_adjoints(double* x) { return x; }
+
+    /**
+     * Within thread reduction.
+     */
+    inline void operator()(const tbb::blocked_range<size_t>& r) {
+      if (r.empty()) {
+        return;
+      }
+
+      if (args_adjoints_.size() == 0) {
+        args_adjoints_ = Eigen::VectorXd::Zero(this->num_shared_terms_);
+      }
+
+      const nested_rev_autodiff begin_nest;
+      // create a deep copy of all var's so that these are not
+      // linked to any outer AD tree
+      std::decay_t<Vec> local_sub_slice;
+      local_sub_slice.reserve(r.size());
+      for (int i = r.begin(); i < r.end(); ++i) {
+        local_sub_slice.emplace_back(deep_copy(vmapped_[i]));
+      }
+      auto args_tuple_local_copy = apply(
+          [&](auto&&... args) {
+            return std::tuple<decltype(deep_copy(args))...>(deep_copy(args)...);
+          },
+          this->args_tuple_);
+      var sub_sum_v = apply(
+          [&](auto&&... args) {
+            return ReduceFunction()(r.begin(), r.end() - 1, local_sub_slice,
+                                    this->msgs_, args...);
+          },
+          args_tuple_local_copy);
+      sub_sum_v.grad();
+      sum_ += sub_sum_v.val();
+      accumulate_adjoints(
+          this->sliced_partials_ + r.begin() * per_job_sliced_terms_,
+          local_sub_slice);
+      apply(
+          [&](auto&&... args) {
+            return accumulate_adjoints(args_adjoints_.data(),
+                                       std::forward<decltype(args)>(args)...);
+          },
+          std::move(args_tuple_local_copy));
+    }
+
+    /**
+     * Join the sum of the reducers
+     * @param child A sub-reducer that is aggregated to make the final reduced
+     * value.
+     * @note side effect of updating `sum_` and
+     *  `args_adjoints_`
+     */
+    void join(const recursive_reducer& rhs) {
+      this->sum_ += rhs.sum_;
+      if (this->args_adjoints_.size() != 0 && rhs.args_adjoints_.size() != 0) {
+        this->args_adjoints_ += rhs.args_adjoints_;
+      } else if (this->args_adjoints_.size() == 0
+                 && rhs.args_adjoints_.size() != 0) {
+        this->args_adjoints_ = rhs.args_adjoints_;
+      }
+    }
+  };
+
+  /**
+   * Count the number of vars in an vector.
+   * @tparam Vec type of standard container holding vars.
+   * @tparam Pargs Types to be forwarded to recursive call of `count_var_impl`.
+   * @param[in] count The current count of the number of vars.
+   * @param[in] x A vector holding vars.
+   * @param[in] args objects to be forwarded to recursive call of
+   * `count_var_impl`
+   */
+  template <typename VecVar, require_std_vector_vt<is_var, VecVar>* = nullptr,
+            typename... Pargs>
+  inline size_t count_var_impl(size_t count, VecVar&& x,
+                               Pargs&&... args) const {
```

Comment:
Could I just rewrite this:

```
inline size_t count_var_impl(size_t count, std::vector<var>&& x, Pargs&&... args)
```

My instinct is to go one step further and do:

```
inline size_t count_var_impl(size_t count, const std::vector<var>& x, Pargs&&... args)
```

But this function exclusively deals with std::vectors filled with vars so I think we don't need the templates.

Reply:
I see why doing that for the Eigen types makes sense (cause they could possibly be Arrays, or Sparse, or some weird Eigen type) though.

Reply:
Oops @SteveBronder so you get a ping

Reply:
`std::vector<var>&&` will only bind to rvalues (temporary values)

Reply:
Any downside to ```const std::vector<var>&```?

Reply:
Not here at all idt count var ever sees an rvalue. I personally just prefer universal references